src/stream/src/executor/aggregation/agg_impl/foldable.rs

```diff
@@ -649,4 +657,354 @@ mod tests {
+        range.shuffle(&mut rand::thread_rng());
+        let shuffled = range.into_iter().collect_vec();
+        for item in shuffled.iter().take(count_ones) {
+            builder.set(*item, true);
+        }
+        builder.finish()
+    }
+
+    fn bench_i64(
+        b: &mut Bencher,
+        mut agg: Box<dyn StreamingAggImpl>,
+        agg_desc: &str,
+        chunk_size: usize,
+        vis_rate: f64,
+        iter_count: usize,
+        append_only: bool,
+    ) {
+        println!(
+            "benching {} agg, chunk_size {}, vis_rate {}, iter_count {}",
+            agg_desc, chunk_size, vis_rate, iter_count
+        );
+        let mut rng = rand::thread_rng();
+        let mut ops: Vec<Op> = vec![];
+        let mut data_builder = I64ArrayBuilder::new(chunk_size);
+        let mut cur_data: Vec<i64> = vec![];
+        let bitmap = if vis_rate < 1.0 {
+            Some(gen_rand_bitmap(
+                chunk_size,
+                (chunk_size as f64 * vis_rate) as usize,
+            ))
+        } else {
+            None
+        };
+
+        if let Some(bitmap) = bitmap.as_ref() {
+            for i in 0..chunk_size {
+                // SAFETY(value_at_unchecked): the idx is always in bound.
+                unsafe {
+                    if bitmap.is_set_unchecked(i) {
+                        let op = if append_only || cur_data.is_empty() || rng.gen() {
+                            Op::Insert
+                        } else {
+                            Op::Delete
+                        };
+                        ops.push(op);
+                        if op == Op::Insert {
+                            let value = rng.gen::<i32>() as i64;
+                            data_builder.append_n(1, Some(value));
+                            cur_data.push(value);
+                        } else {
+                            let idx = rng.gen_range(0..cur_data.len());
+                            data_builder.append_n(1, Some(cur_data[idx]));
+                            cur_data.remove(idx);
+                        }
+                    } else {
+                        ops.push(Op::Insert);
+                        data_builder.append_n(1, Some(1234567890));
+                    }
+                }
+            }
+        } else {
+            for _ in 0..chunk_size {
+                let op = if append_only || cur_data.is_empty() || rng.gen() {
+                    Op::Insert
+                } else {
+                    Op::Delete
+                };
+                ops.push(op);
+                if op == Op::Insert {
+                    let value = rng.gen::<i32>() as i64;
+                    data_builder.append_n(1, Some(value));
+                    cur_data.push(value);
+                } else {
+                    let idx = rng.gen_range(0..cur_data.len());
+                    data_builder.append_n(1, Some(cur_data[idx]));
+                    cur_data.remove(idx);
+                }
+            }
+        }
```

Comment:
Seems like most of this is duplicated with above, could you please refactor this into a `build_data_chunk` function or something similar?

After that, this should become a lot more readable.